src/idns/commands/query.rs

```diff
@@ -0,0 +1,437 @@
+//! The query command of _idns._
+
+use std::collections::HashSet;
+use std::net::{IpAddr, SocketAddr};
+use std::str::FromStr;
+use std::time::Duration;
+use bytes::Bytes;
+use domain::base::iana::Rtype;
+use domain::base::message::Message;
+use domain::base::message_builder::MessageBuilder;
+use domain::base::name::{Name, ToName, UncertainName};
+use domain::net::client;
+use domain::net::client::request::{RequestMessage, SendRequest};
+use domain::rdata::{AllRecordData, Ns, Soa};
+use domain::resolv::StubResolver;
+use domain::resolv::stub;
+use domain::resolv::stub::conf::{ResolvConf, ServerConf, Transport};
+use crate::idns::error::Error;
+use crate::idns::output::OutputFormat;
+
+
+#[derive(Clone, Debug, clap::Args)]
+pub struct Query {
+    /// The name of the resource records to look up
+    #[arg(value_name="QUERY_NAME")]
+    qname: Name<Vec<u8>>,
+
+    /// The record type to look up
+    #[arg(value_name="QUERY_TYPE", default_value = "A")]
+    qtype: Rtype,
+
+    /// The server to send the query to. System servers used if missing
+    #[arg(short, long, value_name="ADDR_OR_HOST")]
+    server: Option<ServerName>,
+
+    /// The port of the server to send query to.
+    #[arg(short = 'p', long = "port", requires = "server")]
+    port: Option<u16>,
+
+    /// Use only IPv4 for communication.
+    #[arg(short = '4', long, conflicts_with = "ipv6")]
+    ipv4: bool,
+
+    /// Use only IPv6 for communication.
+    #[arg(short = '6', long, conflicts_with = "ipv4")]
+    ipv6: bool,
+
+    /// Use only TCP.
+    #[arg(short, long)]
+    tcp: bool,
+
+    /// Use the given message ID. Random if missing.
+    #[arg(long)]
+    id: Option<u16>,
+
+    /// Unset the RD flag in the request.
+    #[arg(long)]
+    no_rd: bool,
+
+    /// Disable all sanity checks.
+    #[arg(long, short = 'f')]
+    force: bool,
+
+    /// Set the timeout for a query.
+    #[arg(long, value_name="SECONDS")]
+    timeout: Option<f32>,
+
+    /// Set the number of retries over UDP.
```

Comment:
Why only over UDP?

Reply:
Retries are only used with UDP since it is an unreliable protocol. TCP and TLS are reliable so retrying is not necessary.

Reply:
A busy server may still timeout however, or be temporarily unreachable?

Reply:
Yes, in that case I want `idns query` to error out.